core/util/http.py

```diff
@@ -265,7 +270,20 @@ def _request_with_timeout(cls, url, make_request_with, *args, **kwargs):
                 # gets added on here. But if you do pass in both
                 # arguments, it will still work.
                 args = args + (url,)
-            response = make_request_with(*args, **kwargs)
+
+            if make_request_with == sessions.Session.request:
+                with sessions.Session() as session:
+                    if max_retry_count is not None:
+                        retry_strategy = Retry(total=max_retry_count)
```

Comment:
There is no explicit information regarding the status codes treated by `Retry` as error status code.
I think it uses the same logic as `requests` library [does](https://docs.python-requests.org/en/master/api/#requests.Response.ok):
> _This attribute checks if the status code of the response is between 400 and 600 to see if there was a client error or a server error. If the status code is between 200 and 400, this will return True. This is not a check to see if the response code is 200 OK._

Basically, it retries on every code between 400 and 600.

Reply:
The [urllib3 documentation](https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html) mentions:
```
status_forcelist (iterable) –

A set of integer HTTP status codes that we should force a retry on. A retry is initiated if the request method is in allowed_methods and the response status code is in status_forcelist.

By default, this is disabled with None.
```

Then mentions:
```
RETRY_AFTER_STATUS_CODES = frozenset({413, 429, 503})

    Default status codes to be used for status_forcelist
```

So I'm not entirely sure what will happen here.

And again, I'm also wondering if there are some "unrecoverable" errors, like not being authenticated/having the wrong credentials.

Reply:
The number of retries is always restricted and we cannot loop forever, so I think we're pretty safe here. Even if some "unrecoverable" error happens CM will fail after finishing the last retry.
Do you have a list of errors when you don't want to retry at all?

Reply:
Okay. Let's leave it as is for now. Would you add a TODO to verify the status codes for which it retries. I don't feel confident about that, but I don't want to hold this up any longer.

Reply:
Tim, sure, I added a TODO.